app/services/action_queue.rb

```diff
@@ -0,0 +1,135 @@
+require 'active_support/concern'
+
+module ActionQueue
+  module Enqueable
+    extend ActiveSupport::Concern
+
+    def initialize(action)
+      @action = action
+    end
+
+    def push
+      ChampaignQueue.push(payload)
+    end
+
+    class_methods do
+      def push(action)
+        new(action).push
+      end
+    end
+  end
+
+  class Pusher
+    def self.push(action)
+      action.donation? ? DonationAction.push(action) : PetitionAction.push(action)
+    end
+  end
+
+  class PetitionAction
+    include Enqueable
+
+    def payload
+      {
+        type: 'action',
+        params: {
+          page: "#{@action.page.slug}-petition"
+        }.merge(@action.form_data)
+      }.deep_symbolize_keys
+    end
+  end
+
+  class DonationAction
+    include Enqueable
+
+    def payload
+      {
+        type:  'donation',
+        params: {
+          donationpage: {
+            name:             "#{@action.page.slug}-donation",
+            payment_account:  get_payment_account
+          },
+          order: {
+            amount:         data[:amount],
+            card_num:       card_num,
+            card_code:      '007',
+            exp_date_month: expire_month,
+            exp_date_year:  expire_year,
+            currency:       data[:currency]
+          },
+          action: {
+            source: data[:source]
+          },
+          user: user_data
+        }
+      }
+    end
+
+    def user_data
+      data.select{ |k, v| k.to_s =~ /^action_/ }.
+        symbolize_keys.merge(
+          first_name: member.first_name,
+          last_name:  member.last_name,
+          email:      member.email,
+          country:    member.country,
+          postal:     data[:postal],
```

Comment:
Yeah this is my only outstanding issue, I'm also confused why only some params receive this treatment, eg `address1` but not `address2`